js/modules/loki_p2p_api.js

```diff
@@ -16,43 +18,80 @@ class LokiP2pAPI extends EventEmitter {
     });
   }
 
-  updateContactP2pDetails(pubKey, address, port, isOnline = false) {
+  updateContactP2pDetails(pubKey, address, port, isPing = false) {
     // Stagger the timers so the friends don't ping each other at the same time
     const timerDuration =
       pubKey < this.ourKey
         ? 60 * 1000 // 1 minute
         : 2 * 60 * 1000; // 2 minutes
 
-    if (this.contactP2pDetails[pubKey]) {
-      clearTimeout(this.contactP2pDetails[pubKey].pingTimer);
+    if (!this.contactP2pDetails[pubKey]) {
+      // We didn't have this contact's details
+      this.contactP2pDetails[pubKey] = {
+        address,
+        port,
+        timerDuration,
+        pingTimer: null,
+        isOnline: false,
+      };
+      // Try ping
+      this.pingContact(pubKey);
+      return;
     }
 
-    this.contactP2pDetails[pubKey] = {
-      address,
-      port,
-      timerDuration,
-      isOnline: false,
-      pingTimer: null,
-    };
+    // We already had this contact's details
+    const baseDetails = { ...this.contactP2pDetails[pubKey] };
 
-    if (isOnline) {
+    if (isPing) {
+      // Received a ping
+      // Update details in case they are new and mark online
+      this.contactP2pDetails[pubKey].address = address;
+      this.contactP2pDetails[pubKey].port = port;
       this.setContactOnline(pubKey);
       return;
     }
 
+    // Received a storage broadcast message
+    if (
+      baseDetails.isOnline ||
+      baseDetails.address !== address ||
+      baseDetails.port !== port
+    ) {
+      // Had the contact marked as online and details we had were the same
+      this.pingContact(pubKey);
+      return;
+    }
+
+    // Had the contact marked as offline or got new details
+    this.contactP2pDetails[pubKey].address = address;
+    this.contactP2pDetails[pubKey].port = port;
+    this.setContactOffline(pubKey);
     this.pingContact(pubKey);
   }
 
   getContactP2pDetails(pubKey) {
     return this.contactP2pDetails[pubKey] || null;
   }
 
+  isContactOnline(pubKey) {
+    const contactDetails = this.contactP2pDetails[pubKey];
+    if (!contactDetails || !contactDetails.isOnline) {
+      return false;
+    }
+    return contactDetails.isOnline;
```

Comment:
```suggestion
    return !!(contactDetails && contactDetails.isOnline);
```
this function can be shortened to that

Reply:
Mmm I don't really like `!!`, it scares and confuses me

Reply:
it just converts any value to `boolean`

Reply:
I agree with Mikunj, we should find a place in our hearts to welcome `!!` in our lives.

Reply:
Ya I mean I get how it works I just think it takes more of my brain power to parse what that one liner is doing when I go back to read the code but maybe will become more natural with repeated use :+1: